File: app/src/main/java/com/example/studybuddy/mlc/MLCLLMService.java
package com.example.studybuddy.mlc;

import android.content.Context;
import android.util.Log;

import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
import org.apache.commons.compress.utils.IOUtils;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Arrays;

/**
 * Service for interacting with the on-device MLC-LLM engine.
 * This class handles model loading, text generation, and resource management.
 */
public class MLCLLMService {
    private static final String TAG = "MLCLLMService";
    private static final String MODEL_DIRECTORY_NAME = "mlc_models";
    private static final String DEFAULT_MODEL_ASSET_PATH = "models/gemma2_2b_it";
    
    private final Context context;
    private boolean isInitialized = false;
    private String modelPath;
    private String lastErrorMessage = null;
    private boolean streamingEnabled = true;
    
    public MLCLLMService(Context context) {
        this.context = context.getApplicationContext();
        Log.d(TAG, "MLCLLMService initialized with context from: " + context.getPackageName());
    }
    
    /**
     * Initialize the LLM with the default model.
     * @return true if initialization was successful
     */
    public boolean initialize() {
        Log.d(TAG, "Initializing with default model path: " + DEFAULT_MODEL_ASSET_PATH);
        return initialize(DEFAULT_MODEL_ASSET_PATH);
    }
    
    /**
     * Get the last error message if initialization or generation failed
     */
    public String getLastErrorMessage() {
        return lastErrorMessage != null ? lastErrorMessage : "Unknown error";
    }
    
    /**
     * Initialize the LLM with a specified model.
     * @param modelAssetPath Path to the model directory in assets
     * @return true if initialization was successful
     */
    public boolean initialize(String modelAssetPath) {
        if (isInitialized) {
            Log.i(TAG, "LLM already initialized");
            return true;
        }
        
        // Reset error message
        lastErrorMessage = null;
        
        // Print all available libraries
        try {
            File jniLibsDir = new File(context.getApplicationInfo().nativeLibraryDir);
            Log.d(TAG, "JNI libraries directory: " + jniLibsDir.getAbsolutePath());
            if (jniLibsDir.exists()) {
                String[] libraries = jniLibsDir.list();
                if (libraries != null) {
                    Log.d(TAG, "Available libraries: " + Arrays.toString(libraries));
                } else {
                    Log.w(TAG, "No libraries found in native directory");
                }
            } else {
                Log.w(TAG, "Native library directory does not exist");
            }
        } catch (Exception e) {
            Log.e(TAG, "Error checking native libraries: " + e.getMessage(), e);
        }
        
        // Check if libraries are loaded
        if (!TVMBridge.areLibrariesLoaded()) {
            lastErrorMessage = "Failed to load MLC-LLM native libraries";
            Log.e(TAG, lastErrorMessage);
            return false;
        }
        
        try {
            // Check if model files already exist in the internal storage
            String modelName = modelAssetPath.substring(modelAssetPath.lastIndexOf('/') + 1);
            File modelDir = new File(context.getFilesDir(), MODEL_DIRECTORY_NAME);
            File specificModelDir = new File(modelDir, modelName);
            
            if (specificModelDir.exists() && specificModelDir.isDirectory()) {
                String[] files = specificModelDir.list();
                if (files != null && files.length > 0) {
                    Log.i(TAG, "Found existing model files in " + specificModelDir.getAbsolutePath());
                    Log.d(TAG, "Files: " + Arrays.toString(files));
                    
                    // Check if config.json exists
                    if (Arrays.asList(files).contains("config.json")) {
                        Log.i(TAG, "Using existing model files");
                        modelPath = specificModelDir.getAbsolutePath();
                        
                        // Initialize the MLC-LLM runtime with existing files
                        isInitialized = TVMBridge.initRuntime(modelPath);
                        
                        if (isInitialized) {
                            Log.i(TAG, "MLC-LLM initialized successfully with existing model files");
                            // Set default parameters
                            TVMBridge.setTemperature(0.7f);
                            TVMBridge.setTopP(0.95f);
                            TVMBridge.setRepetitionPenalty(1.1f);
                            return true;
                        } else {
                            Log.w(TAG, "Failed to initialize with existing model files, will try to extract again");
                        }
                    }
                }
            }
            
            // Print assets to debug
            String[] assetFiles = context.getAssets().list("");
            Log.d(TAG, "Assets root: " + Arrays.toString(assetFiles));
            
            // Also check if pre-extracted model exists in assets
            try {
                String[] preExtractedFiles = context.getAssets().list(modelAssetPath);
                if (preExtractedFiles != null && preExtractedFiles.length > 0) {
                    Log.d(TAG, "Found pre-extracted model in assets with " + preExtractedFiles.length + " files");
                    
                    // Check if config.json exists among pre-extracted files
                    boolean hasConfigJson = false;
                    for (String file : preExtractedFiles) {
                        if ("config.json".equals(file)) {
                            hasConfigJson = true;
                            break;
                        }
                    }
                    
                    if (hasConfigJson) {
                        Log.i(TAG, "Using pre-extracted model files from assets");
                    }
                }
            } catch (IOException e) {
                Log.d(TAG, "No pre-extracted model found in assets: " + e.getMessage());
            }
            
            // Check models directory in assets
            String[] modelAssets = context.getAssets().list("models");
            Log.d(TAG, "Models in assets: " + Arrays.toString(modelAssets));
            
            // Extract model files to accessible storage
            Log.i(TAG, "Extracting model files from assets: " + modelAssetPath);
            File extractedModelDir = extractModelFromAssets(modelAssetPath);
            
            if (extractedModelDir == null) {
                lastErrorMessage = "Failed to extract model files from assets";
                Log.e(TAG, lastErrorMessage);
                
                // Try to extract the tar file instead
                if (modelAssetPath.equals(DEFAULT_MODEL_ASSET_PATH)) {
                    Log.i(TAG, "Attempting to extract from tar file...");
                    extractedModelDir = extractModelFromTarFile();
                    
                    if (extractedModelDir == null) {
                        lastErrorMessage = "Failed to extract model from tar file";
                        Log.e(TAG, lastErrorMessage);
                        return false;
                    }
                } else {
                    return false;
                }
            }
            
            // List extracted files for debugging
            Log.d(TAG, "Model directory contents: " + Arrays.toString(extractedModelDir.list()));
            
            // Initialize the MLC-LLM runtime
            modelPath = extractedModelDir.getAbsolutePath();
            Log.i(TAG, "Initializing MLC-LLM with model at: " + modelPath);
            isInitialized = TVMBridge.initRuntime(modelPath);
            
            if (isInitialized) {
                Log.i(TAG, "MLC-LLM initialized successfully");
                // Set default parameters
                TVMBridge.setTemperature(0.7f);
                TVMBridge.setTopP(0.95f);
                TVMBridge.setRepetitionPenalty(1.1f);
            } else {
                lastErrorMessage = "TVMBridge.initRuntime() returned false";
                Log.e(TAG, lastErrorMessage);
            }
            
            return isInitialized;
        } catch (Exception e) {
            lastErrorMessage = "Error initializing MLC-LLM: " + e.getMessage();
            Log.e(TAG, lastErrorMessage, e);
            return false;
        }
    }
    
    /**
     * Extract model from a tar file in assets using Apache Commons Compress
     */
    private File extractModelFromTarFile() {
        InputStream inputStream = null;
        TarArchiveInputStream tarIn = null;
        
        try {
            Log.d(TAG, "Extracting model from tar file");
            // Look for gemma tar file
            String tarFileName = "gemma-2b-it-q4f16_1-android.tar";
            
            // Create a model directory
            File modelDir = new File(context.getFilesDir(), MODEL_DIRECTORY_NAME);
            if (!modelDir.exists()) {
                if (!modelDir.mkdirs()) {
                    Log.e(TAG, "Failed to create model directory: " + modelDir.getAbsolutePath());
                    return null;
                }
            }
            
            // Create a subdirectory for this specific model
            String modelName = "gemma2_2b_it";
            File specificModelDir = new File(modelDir, modelName);
            if (!specificModelDir.exists()) {
                if (!specificModelDir.mkdirs()) {
                    Log.e(TAG, "Failed to create specific model directory: " + specificModelDir.getAbsolutePath());
                    return null;
                }
            }
            
            // Open the tar file from assets
            inputStream = context.getAssets().open(tarFileName);
            BufferedInputStream bufferedInputStream = new BufferedInputStream(inputStream);
            tarIn = new TarArchiveInputStream(bufferedInputStream);
            
            TarArchiveEntry entry;
            int extractedFiles = 0;
            
            Log.d(TAG, "Starting extraction of tar entries to " + specificModelDir.getAbsolutePath());
            while ((entry = tarIn.getNextTarEntry()) != null) {
                if (entry.isDirectory()) {
                    File directory = new File(specificModelDir, entry.getName());
                    if (!directory.exists() && !directory.mkdirs()) {
                        Log.w(TAG, "Failed to create directory: " + directory.getAbsolutePath());
                    }
                    continue;
                }
                
                // Fix file path (remove directory prefix if present)
                String filename = entry.getName();
                int lastSlashIndex = filename.lastIndexOf('/');
                if (lastSlashIndex != -1) {
                    filename = filename.substring(lastSlashIndex + 1);
                }
                
                File outFile = new File(specificModelDir, filename);
                
                // Skip if file already exists
                if (outFile.exists() && outFile.length() > 0) {
                    Log.d(TAG, "File already exists, skipping: " + filename);
                    continue;
                }
                
                // Ensure parent directory exists
                File parent = outFile.getParentFile();
                if (parent != null && !parent.exists()) {
                    parent.mkdirs();
                }
                
                Log.d(TAG, "Extracting: " + filename + " (size: " + entry.getSize() + " bytes)");
                
                FileOutputStream fos = new FileOutputStream(outFile);
                IOUtils.copy(tarIn, fos);
                fos.close();
                extractedFiles++;
            }
            
            // Create a config.json file if it doesn't exist
            File configFile = new File(specificModelDir, "config.json");
            if (!configFile.exists()) {
                Log.d(TAG, "Creating minimal config.json file");
                FileOutputStream fos = new FileOutputStream(configFile);
                String configContent = "{\n" +
                    "    \"model_name\": \"gemma-2b-it\",\n" +
                    "    \"quantization\": \"q4f16_1\",\n" +
                    "    \"model_lib\": \"libmlc_llm.so\",\n" +
                    "    \"runtime_lib\": \"libtvm_runtime.so\"\n" +
                    "}";
                fos.write(configContent.getBytes());
                fos.close();
                extractedFiles++;
            }
            
            Log.i(TAG, "Extracted " + extractedFiles + " files from tar");
            
            // Check if extraction worked
            if (extractedFiles > 0) {
                Log.i(TAG, "Successfully extracted files to " + specificModelDir.getAbsolutePath());
                return specificModelDir;
            } else {
                Log.e(TAG, "No files were extracted from the tar file");
                return null;
            }
        } catch (Exception e) {
            Log.e(TAG, "Error extracting from tar file: " + e.getMessage(), e);
            return null;
        } finally {
            try {
                if (tarIn != null) tarIn.close();
                if (inputStream != null) inputStream.close();
            } catch (IOException e) {
                Log.e(TAG, "Error closing streams: " + e.getMessage(), e);
            }
        }
    }
    
    /**
     * Generate text based on the provided prompt.
     * @param prompt The input prompt to generate text from
     * @param maxTokens Maximum number of tokens to generate
     * @return Generated text response
     */
    public String generateText(String prompt, int maxTokens) {
        if (!isInitialized) {
            Log.e(TAG, "MLC-LLM not initialized. Call initialize() first.");
            return "Error: MLC-LLM not initialized";
        }
        
        try {
            Log.i(TAG, "Generating text for prompt: " + prompt);
            String response = TVMBridge.generateText(prompt, maxTokens);
            Log.i(TAG, "Text generation completed");
            return response;
        } catch (Exception e) {
            Log.e(TAG, "Error generating text: " + e.getMessage(), e);
            return "Error generating text: " + e.getMessage();
        }
    }
    
    /**
     * Adjust the temperature parameter for text generation.
     * Higher values increase randomness.
     * @param temperature Value between 0.0 and 1.0
     */
    public void setTemperature(float temperature) {
        if (isInitialized) {
            TVMBridge.setTemperature(temperature);
        }
    }
    
    /**
     * Adjust the top-p (nucleus sampling) parameter.
     * @param topP Value between 0.0 and 1.0
     */
    public void setTopP(float topP) {
        if (isInitialized) {
            TVMBridge.setTopP(topP);
        }
    }
    
    /**
     * Adjust the repetition penalty to reduce repetitive text.
     * @param penalty Usually between 1.0 and 1.5
     */
    public void setRepetitionPenalty(float penalty) {
        if (isInitialized) {
            TVMBridge.setRepetitionPenalty(penalty);
        }
    }
    
    /**
     * Clean up resources used by the LLM.
     */
    public void shutdown() {
        if (isInitialized) {
            TVMBridge.destroyRuntime();
            isInitialized = false;
            Log.i(TAG, "MLC-LLM resources released");
        }
    }
    
    /**
     * Extract model files from assets to accessible storage.
     * @param assetPath Path to the model in assets
     * @return The directory containing the extracted model
     */
    private File extractModelFromAssets(String assetPath) {
        try {
            // Create the model directory in app's files directory
            File modelDir = new File(context.getFilesDir(), MODEL_DIRECTORY_NAME);
            if (!modelDir.exists()) {
                modelDir.mkdirs();
            }
            
            // Create a subdirectory for this specific model
            String modelName = assetPath.substring(assetPath.lastIndexOf('/') + 1);
            File specificModelDir = new File(modelDir, modelName);
            if (!specificModelDir.exists()) {
                specificModelDir.mkdirs();
            }
            
            // Check if the asset directory exists
            try {
                context.getAssets().open(assetPath); // This will throw an exception if it's a directory
                Log.e(TAG, "Expected directory but found file at: " + assetPath);
                return null;
            } catch (IOException e) {
                // This is expected for directories
            }
            
            // Get the list of files in the asset directory
            String[] files;
            try {
                files = context.getAssets().list(assetPath);
            } catch (IOException e) {
                Log.e(TAG, "Failed to list assets in path: " + assetPath + ", error: " + e.getMessage());
                return null;
            }
            
            if (files != null && files.length > 0) {
                Log.d(TAG, "Found " + files.length + " files in asset path: " + assetPath);
                
                for (String fileName : files) {
                    Log.d(TAG, "Processing asset file: " + fileName);
                    
                    // Copy each file to the model directory
                    try {
                        InputStream input = context.getAssets().open(assetPath + "/" + fileName);
                        File outFile = new File(specificModelDir, fileName);
                        
                        // Skip if file already exists
                        if (outFile.exists() && outFile.length() > 0) {
                            Log.d(TAG, "File already exists, skipping: " + fileName);
                            continue;
                        }
                        
                        OutputStream output = new FileOutputStream(outFile);
                        byte[] buffer = new byte[8192];
                        int read;
                        long total = 0;
                        while ((read = input.read(buffer)) != -1) {
                            output.write(buffer, 0, read);
                            total += read;
                        }
                        input.close();
                        output.flush();
                        output.close();
                        Log.d(TAG, "Extracted file: " + fileName + " (" + total + " bytes)");
                    } catch (IOException e) {
                        Log.e(TAG, "Error extracting file " + fileName + ": " + e.getMessage());
                    }
                }
                
                // Check if we actually extracted any files
                if (specificModelDir.list() != null && specificModelDir.list().length > 0) {
                    Log.i(TAG, "Successfully extracted model files to: " + specificModelDir.getAbsolutePath());
                    return specificModelDir;
                } else {
                    Log.e(TAG, "No files were extracted to: " + specificModelDir.getAbsolutePath());
                    return null;
                }
            } else {
                Log.e(TAG, "No files found in asset path: " + assetPath);
                return null;
            }
        } catch (Exception e) {
            Log.e(TAG, "Error extracting model files: " + e.getMessage(), e);
            return null;
        }
    }
    
    /**
     * Generate text using streaming mode
     * @param prompt Input prompt
     * @param maxTokens Maximum tokens to generate
     * @param callback Callback to receive generated tokens
     * @return True if streaming generation started successfully
     */
    public boolean generateTextStreaming(String prompt, int maxTokens, TVMBridge.StreamingCallback callback) {
        if (!isInitialized) {
            Log.e(TAG, "Cannot generate text: model not initialized");
            lastErrorMessage = "Model not initialized";
            return false;
        }
        
        try {
            Log.d(TAG, "Starting streaming text generation for prompt: " + prompt);
            return TVMBridge.startStreamingGeneration(prompt, maxTokens, callback);
        } catch (Exception e) {
            lastErrorMessage = "Error in streaming text generation: " + e.getMessage();
            Log.e(TAG, lastErrorMessage, e);
            return false;
        }
    }
    
    /**
     * Stop any ongoing streaming text generation
     */
    public void stopTextStreaming() {
        try {
            TVMBridge.stopStreamingGeneration();
        } catch (Exception e) {
            Log.e(TAG, "Error stopping streaming generation: " + e.getMessage(), e);
        }
    }
    
    /**
     * Check if streaming text generation is enabled
     * @return true if streaming is enabled
     */
    public boolean isStreamingEnabled() {
        return streamingEnabled;
    }
    
    /**
     * Enable or disable streaming text generation
     * @param enabled true to enable streaming, false to disable
     */
    public void setStreamingEnabled(boolean enabled) {
        streamingEnabled = enabled;
    }
    
    /**
     * Generate text using the most appropriate method (streaming or non-streaming)
     * @param prompt Input prompt
     * @param maxTokens Maximum tokens to generate
     * @param callback Callback for streaming tokens (if streaming is enabled)
     * @return Generated text if not using streaming, empty string if using streaming
     */
    public String generateTextAuto(String prompt, int maxTokens, TVMBridge.StreamingCallback callback) {
        if (streamingEnabled && callback != null) {
            generateTextStreaming(prompt, maxTokens, callback);
            return ""; // Response will come through callback
        } else {
            return generateText(prompt, maxTokens);
        }
    }
} 

File: app/src/main/java/com/example/studybuddy/ml/MLCLLMService.kt
package com.example.studybuddy.ml

import android.content.Context
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.withContext
import java.io.File
import java.io.IOException

/**
 * A service for managing MLC-LLM model loading and inference
 */
class MLCLLMService private constructor(private val context: Context) {
    companion object {
        private const val TAG = "MLCLLMService"
        private const val DEFAULT_MODEL_PATH = "models/gemma2_2b_it"
        private const val LIB_DIR = "lib"
        private const val APP_ASSETS_MODEL_DIR = "models/gemma2_2b_it"
        private const val MODEL_DIR = "models/gemma2_2b_it"
        private const val CONFIG_FILE = "config.json"
        
        @Volatile
        private var instance: MLCLLMService? = null
        
        fun getInstance(context: Context): MLCLLMService {
            return instance ?: synchronized(this) {
                instance ?: MLCLLMService(context.applicationContext).also { 
                    instance = it
                    Log.d(TAG, "MLCLLMService initialized with context from: ${context.packageName}")
                }
            }
        }
    }
    
    private val _isModelLoaded = MutableStateFlow(false)
    val isModelLoaded: StateFlow<Boolean> = _isModelLoaded
    
    private val _textStream = MutableStateFlow<List<String>>(emptyList())
    val textStream: StateFlow<List<String>> = _textStream
    
    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage
    
    private var lastStreamedText = ""
    
    init {
        // Initialize the model on first use
        _isModelLoaded.value = false
        _errorMessage.value = null
    }
    
    /**
     * Load the MLC-LLM model from assets
     * @throws RuntimeException if the model fails to load
     */
    suspend fun loadModel(): Boolean = withContext(Dispatchers.IO) {
        try {
            Log.d(TAG, "Initializing model...")
            _errorMessage.value = null
            
            // List available libraries
            val libDir = File(context.applicationInfo.nativeLibraryDir)
            Log.d(TAG, "JNI libraries directory: ${libDir.absolutePath}")
            if (libDir.exists()) {
                val libs = libDir.listFiles()?.map { it.name } ?: emptyList()
                Log.d(TAG, "Available libraries: $libs")
            }
            
            // First, check what's in the assets directory
            val assetsList = context.assets.list("")?.toList() ?: emptyList()
            Log.d(TAG, "Assets root directory contents: $assetsList")
            
            if ("models" in assetsList) {
                val modelsContents = context.assets.list("models")?.toList() ?: emptyList()
                Log.d(TAG, "Assets models directory contents: $modelsContents")
                
                if ("gemma2_2b_it" in modelsContents) {
                    val modelContents = context.assets.list("models/gemma2_2b_it")?.toList() ?: emptyList()
                    Log.d(TAG, "Assets model directory contents: $modelContents")
                }
            }
            
            // Create the model directory in app private storage
            val appModelDir = File(context.filesDir, DEFAULT_MODEL_PATH)
            if (!appModelDir.exists()) {
                val created = appModelDir.mkdirs()
                Log.d(TAG, "Created model directory: $created at ${appModelDir.absolutePath}")
            } else {
                val files = appModelDir.listFiles()?.map { it.name } ?: emptyList()
                Log.d(TAG, "Existing model directory contains: $files")
            }
            
            // Create lib directory
            val libDirPath = File(appModelDir, LIB_DIR)
            if (!libDirPath.exists()) {
                val created = libDirPath.mkdirs()
                Log.d(TAG, "Created lib directory: $created at ${libDirPath.absolutePath}")
            }
            
            // Create the stub config file if it doesn't exist
            val configFile = File(appModelDir, "config.json")
            if (!configFile.exists()) {
                try {
                    configFile.writeText(
                        """
                        {
                            "model_name": "gemma-2b-it",
                            "quantization": "q4f16_1",
                            "model_lib": "libmlc_llm.so",
                            "runtime_lib": "libtvm_runtime.so"
                        }
                        """.trimIndent()
                    )
                    Log.d(TAG, "Created config.json successfully")
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to create config.json: ${e.message}")
                    throw e
                }
            }
            
            // Create stub lib file if it doesn't exist
            val stubLibFile = File(libDirPath, "libgemma-2b-it-q4f16_1.so")
            if (!stubLibFile.exists()) {
                try {
                    stubLibFile.createNewFile()
                    Log.d(TAG, "Created stub lib file successfully")
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to create stub lib file: ${e.message}")
                    throw e
                }
            }
            
            // Copy parameter files from assets
            try {
                // Check if we have parameter files in assets
                val assetsModelDir = "models/gemma2_2b_it"
                if (context.assets.list(assetsModelDir)?.any { it.startsWith("params_shard_") } == true) {
                    Log.d(TAG, "Found parameter files in assets, copying them")
                    
                    val paramFiles = context.assets.list(assetsModelDir)
                        ?.filter { filename: String -> filename.startsWith("params_shard_") } ?: emptyList<String>()
                    
                    for (paramFile in paramFiles) {
                        val destFile = File(appModelDir, paramFile)
                        if (!destFile.exists()) {
                            try {
                                context.assets.open("$assetsModelDir/$paramFile").use { input ->
                                    destFile.outputStream().use { output ->
                                        input.copyTo(output)
                                    }
                                }
                                Log.d(TAG, "Copied $paramFile successfully")
                            } catch (e: Exception) {
                                Log.e(TAG, "Failed to copy $paramFile: ${e.message}")
                            }
                        }
                    }
                } else {
                    Log.d(TAG, "No parameter files found in assets")
                    
                    // If no parameter files in assets, create empty parameter files for testing
                    for (i in 0 until 38) {
                        val paramFile = File(appModelDir, "params_shard_$i.bin")
                        if (!paramFile.exists()) {
                            try {
                                paramFile.createNewFile()
                                Log.d(TAG, "Created empty parameter file: params_shard_$i.bin")
                            } catch (e: Exception) {
                                Log.e(TAG, "Failed to create empty parameter file: ${e.message}")
                            }
                        }
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error copying parameter files: ${e.message}")
                // Continue anyway - we'll still try to initialize
            }
            
            // Also copy tokenizer files if they exist
            try {
                val assetsModelDir = "models/gemma2_2b_it"
                val tokenizerFiles = arrayOf("tokenizer.json", "tokenizer_config.json", "tokenizer.model")
                
                for (tokenizerFile in tokenizerFiles) {
                    val destFile = File(appModelDir, tokenizerFile)
                    if (!destFile.exists()) {
                        try {
                            if (context.assets.list(assetsModelDir)?.contains(tokenizerFile) == true) {
                                context.assets.open("$assetsModelDir/$tokenizerFile").use { input ->
                                    destFile.outputStream().use { output ->
                                        input.copyTo(output)
                                    }
                                }
                                Log.d(TAG, "Copied $tokenizerFile successfully")
                            } else {
                                // Create empty tokenizer file
                                destFile.createNewFile()
                                Log.d(TAG, "Created empty tokenizer file: $tokenizerFile")
                            }
                        } catch (e: Exception) {
                            Log.e(TAG, "Failed to handle $tokenizerFile: ${e.message}")
                        }
                    }
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error handling tokenizer files: ${e.message}")
                // Continue anyway
            }
            
            // Check that we have all the necessary files
            val requiredFiles = listOf("config.json")
            val missingFiles = requiredFiles.filter { !File(appModelDir, it).exists() }
            
            if (missingFiles.isNotEmpty()) {
                val errorMsg = "Missing required files: $missingFiles"
                Log.e(TAG, errorMsg)
                _errorMessage.value = errorMsg
                _isModelLoaded.value = false
                throw RuntimeException(errorMsg)
            }
            
            // Load the model using TVMBridge
            try {
                val tvmBridge = TVMBridge()
                val success = tvmBridge.initializeEngine(appModelDir.absolutePath)
                
                if (success) {
                    Log.i(TAG, "MLC-LLM initialized successfully")
                    _isModelLoaded.value = true
                    _errorMessage.value = null
                } else {
                    val errorMsg = "Failed to initialize MLC-LLM engine"
                    Log.e(TAG, errorMsg)
                    _errorMessage.value = errorMsg
                    _isModelLoaded.value = false
                    throw RuntimeException(errorMsg)
                }
            } catch (e: Exception) {
                val errorMsg = "Error initializing MLC-LLM engine: ${e.message}"
                Log.e(TAG, errorMsg, e)
                _errorMessage.value = errorMsg
                _isModelLoaded.value = false
                throw e
            }
            
            return@withContext _isModelLoaded.value
        } catch (e: Exception) {
            val errorMsg = "Error loading MLC-LLM model: ${e.message}"
            Log.e(TAG, errorMsg, e)
            _errorMessage.value = errorMsg
            _isModelLoaded.value = false
            throw e
        }
    }
    
    /**
     * Generate text with the LLM model
     * @throws RuntimeException if text generation fails
     */
    suspend fun generateText(prompt: String): String = withContext(Dispatchers.IO) {
        if (!_isModelLoaded.value) {
            try {
                loadModel()
            } catch (e: Exception) {
                throw RuntimeException("Failed to load model: ${e.message}")
            }
        }
        
        try {
            val tvmBridge = TVMBridge()
            val response = tvmBridge.chat(prompt)
            return@withContext response
        } catch (e: Exception) {
            val errorMsg = "Error generating text: ${e.message}"
            Log.e(TAG, errorMsg, e)
            _errorMessage.value = errorMsg
            throw RuntimeException(errorMsg)
        }
    }
    
    /**
     * Stream text generation with the LLM model
     * @throws RuntimeException if streaming fails
     */
    suspend fun streamText(prompt: String) = withContext(Dispatchers.IO) {
        // Reset the stream
        _textStream.value = emptyList()
        lastStreamedText = ""
        _errorMessage.value = null
        
        Log.d(TAG, "Starting streaming text generation for prompt: $prompt")
        
        if (!_isModelLoaded.value) {
            try {
                loadModel()
            } catch (e: Exception) {
                val errorMsg = "Failed to load model: ${e.message}"
                _errorMessage.value = errorMsg
                _textStream.value = listOf(errorMsg)
                throw RuntimeException(errorMsg)
            }
        }
        
        try {
            // Use the TVMBridge for streaming text generation
            val tvmBridge = TVMBridge()
            
            // Start streaming in a separate thread
            val thread = Thread {
                try {
                    tvmBridge.streamChat(prompt) { token ->
                        lastStreamedText += token
                        _textStream.value = listOf(lastStreamedText)
                    }
                } catch (e: Exception) {
                    val errorMsg = "Error in streaming thread: ${e.message}"
                    Log.e(TAG, errorMsg, e)
                    _errorMessage.value = errorMsg
                    _textStream.value = listOf(lastStreamedText + "\n\nError: ${e.message}")
                    throw RuntimeException(errorMsg)
                }
            }
            
            thread.start()
            thread.join() // Wait for streaming to complete
            
        } catch (e: Exception) {
            val errorMsg = "Error streaming text: ${e.message}"
            Log.e(TAG, errorMsg, e)
            _errorMessage.value = errorMsg
            _textStream.value = listOf(errorMsg)
            throw RuntimeException(errorMsg)
        }
    }

    /**
     * Check if model files are already available
     */
    suspend fun areModelFilesAvailable(): Boolean = withContext(Dispatchers.IO) {
        // First check if the models folder in assets has the required files
        val assetDir = File(context.filesDir, APP_ASSETS_MODEL_DIR)
        if (assetDir.exists()) {
            val configFile = File(assetDir, "config.json")
            if (configFile.exists()) {
                Log.d(TAG, "Found model files in app assets: $assetDir")
                return@withContext true
            }
        }
        
        // Otherwise check the downloaded model dir
        val modelDir = File(context.filesDir, MODEL_DIR)
        if (!modelDir.exists()) return@withContext false
        
        // Check for config file - only this is absolutely required
        val configFile = File(modelDir, CONFIG_FILE)
        if (!configFile.exists()) {
            return@withContext false
        }
        
        // Check for at least one parameter shard
        val paramFiles = modelDir.listFiles { file -> file.name.startsWith("params_shard_") }
        if (paramFiles == null || paramFiles.isEmpty()) {
            Log.d(TAG, "No parameter shards found")
            return@withContext false
        }
        
        Log.d(TAG, "All model files are available")
        return@withContext true
    }
} 